connector/connect/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectStreamHandler.scala

```diff
@@ -114,10 +123,97 @@ class SparkConnectStreamHandler(responseObserver: StreamObserver[Response]) exte
       responseObserver.onNext(response.build())
     }
 
-    responseObserver.onNext(sendMetricsToResponse(clientId, rows))
+    responseObserver.onNext(sendMetricsToResponse(clientId, dataframe))
     responseObserver.onCompleted()
   }
 
+  def processRowsAsArrowBatches(clientId: String, dataframe: DataFrame): Unit = {
+    val spark = dataframe.sparkSession
+    val schema = dataframe.schema
+    // TODO: control the batch size instead of max records
+    val maxRecordsPerBatch = spark.sessionState.conf.arrowMaxRecordsPerBatch
+    val timeZoneId = spark.sessionState.conf.sessionLocalTimeZone
+
+    SQLExecution.withNewExecutionId(dataframe.queryExecution, Some("collectArrow")) {
+      val rows = dataframe.queryExecution.executedPlan.execute()
+      val numPartitions = rows.getNumPartitions
+      var numSent = 0
+
+      if (numPartitions > 0) {
+        type Batch = (Array[Byte], Long, Long)
+
+        val batches = rows.mapPartitionsInternal { iter =>
+          ArrowConverters
+            .toArrowBatchIterator(iter, schema, maxRecordsPerBatch, timeZoneId)
+        }
+
+        val signal = new Object
+        val partitions = Array.fill[Array[Batch]](numPartitions)(null)
+
+        val processPartition = (iter: Iterator[Batch]) => iter.toArray
+
+        val resultHandler = (partitionId: Int, partition: Array[Batch]) => {
+          signal.synchronized {
+            partitions(partitionId) = partition
+            signal.notify()
+          }
+          val i = 0 // Unit
+        }
+
+        spark.sparkContext.runJob(batches, processPartition, resultHandler)
+
+        var currentPartitionId = 0
+        while (currentPartitionId < numPartitions) {
+          val partition = signal.synchronized {
+            while (partitions(currentPartitionId) == null) {
+              signal.wait()
```

Comment:
Is it to wait the partitions to be fetched in order? I think we can just fetch all and send the first if that arrives. To optimize this, I think we should eventually do the reordering in some way to match with PySpark's implementation. Even we should deduplicate the codes ideally.

Reply:
no, partitions can be fetched by random order. here wait for the `currentPartitionId`-th (start from 0) partition

Reply:
Sorry I am still bit confused on how does this maintain the ordering of partitions. I assume other people when they reading this code, they might be confused as well.

Is it possible to have some developer comment here to explain the algorithm?

Reply:
If the first partition arrives last, the whole dataset stays in the driver's memory, right?

Reply:
Yes. We can look into spilling to deal with these situations.

Reply:
How do you think this way? https://github.com/apache/spark/pull/38468#discussion_r1016118342

Reply:
That is an optimization we can go for after merging this. My only problem with that is that it might scare users, because collecting a dataframe multiple times might have a results that look differently.

Reply:
In reduce phase, the task fetches map data of each partition in random order, w/o local sort, user still sees indeterminate data even the driver returns data by partition number.

Reply:
> If the first partition arrives last, the whole dataset stays in the driver's memory, right?

yes, but at least it's not worse than existing `collect` which always keep whole dataset in memory.

receiving the partitions by order may make it easier to consume in the client, if ordering matters.

I think we will optimize it further, it is just an initial implementation.

Reply:
The current approach still has the risk of holding all the results in the driver memory (assuming the first partition comes last), which violates the design goal of Spark Connect.

I think the Spark driver should send whichever partition that arrives to the client, and the client should allocate an array to hold arrow batches of all partitions. The client need to keep all the result in-memory anyway, so it's better to ask the client to buffer the results ad reorder them by partition id.

Reply:
That makes the client and the API more complicated. I don't want the implementors of the clients to deal with this. We can add an optimization for a dataframe that is unordered in a follow-up, but for now let's just merge the things that works.

Reply:
Agreed, it's also important to keep client implementations simple. This "async collect" should be OK in most cases.